Add initial support for Tornado 16X SQ air conditioner (0x4e2a)

enosh opened the pull request:
# Proposed changes

Add support for Tornado 16X SQ air conditioner (sold in Israel [link](https://www.tornado-top.co.il/catalog/product.aspx?pid=297)). I think it'll also work for related models. Their app is based on AC Freedom so it might work with some models from that brand also. There are features like sleep curves, timers and ambient temperature I didn't implement.

felipediel commented:
Thanks for the contribution. If you try to control it with the hysen class, does it work?

Enosh commented:
It's the first thing I tried, but it fails, the payloads are different in size, info and don't use a CRC for checksum. Figuring out the values in the payload took me the most time, your broadlink-hacktools was invaluable.

felipediel commented on broadlink/climate.py:
```diff
@@ -201,3 +201,282 @@ def set_schedule(self, weekday: List[dict], weekend: List[dict]) -> None:
+        result = target - (diff & 0xffff)
+        return [result & 0xff, (result >> 8) & 0xff]
+
+    def _send_short_payload(self, a):
```
Don't use "a" for variable names. Use something meaningful.

felipediel commented on broadlink/climate.py:
```diff
@@ -201,3 +201,282 @@ def set_schedule(self, weekday: List[dict], weekend: List[dict]) -> None:
+        payload = self.decrypt(bytes(response[0x38:]))
+        return payload
+    
+    def _calculate_checksum(self, arr, target=0x20017, checksum=None):
+        """Calculate checksum of given array,
+        by adding little endian words and subtracting from target.
+        
+        Args:
+            arr (list): 
+            checksum (Optional[int]):  a received checksum, can be used to calc target
+        """
+        diff = 0
+        for i in range(len(arr)): # might need an override
+            if (i % 2 == 0):
+                try:
+                    diff += (arr[i] + (arr[i+1] << 8))
+                except:
+                    # odd array size, add little value
+                    diff += arr[i]
+        
+        if (checksum):
+            print('checksum diff is {}'.format(diff + checksum))
+
+        result = target - (diff & 0xffff)
+        return [result & 0xff, (result >> 8) & 0xff]
```
This function is not necessary. We use sum() for checksums.

Enosh replied:
I rewrote it to use sum(), it's more complex than the other uses of sum() and I think having it as function is more readable. What do you think?

felipediel replied:
How about [this](https://github.com/mjg59/python-broadlink/pull/430#discussion_r494007259)?

felipediel replied:
How about [this](https://github.com/mjg59/python-broadlink/pull/430#discussion_r494007259)?

felipediel commented on broadlink/climate.py:
```diff
@@ -201,3 +201,282 @@ def set_schedule(self, weekday: List[dict], weekend: List[dict]) -> None:
+
+        return data
+
+    def set_advanced(self, state=None, mode=None, set_temp=None, speed=None, swing_v=None, swing_h=None, sleep=None, display=None, health=None):
```
Please use function annotations to make more clear what is the expected input and output.

felipediel commented on broadlink/climate.py:
```diff
@@ -201,3 +201,282 @@ def set_schedule(self, weekday: List[dict], weekend: List[dict]) -> None:
+                display (bool):
+                health (bool):
+        """
+        GET_STATE_CHECKSUM_TARGET = 0x20017
```
Do not start consts with GET_ or SET_. These are function prefixes.

felipediel commented on broadlink/climate.py:
```diff
@@ -201,3 +201,282 @@ def set_schedule(self, weekday: List[dict], weekend: List[dict]) -> None:
+        response = self.send_packet(0x6a, bytearray(payload))
+        check_error(response[0x22:0x24])
+        response_payload = self._decode(response)
+        return response_payload
```
Add a new line to the end of the file.

felipediel commented on broadlink/climate.py:
```diff
@@ -201,3 +201,282 @@ def set_schedule(self, weekday: List[dict], weekend: List[dict]) -> None:
+        check_error(response[0x22:0x24])
+        return (self._decode(response))
+
+    def _send_short_payload_d0_07(self, payload=None):
+        """Not sure what this payload does"""
+        packet = bytearray(16)
+        packet[0x00] = 0xd0
+        packet[0x01] = 0x07
+
+        response = self.send_packet(0x6a, packet)
+        check_error(response[0x22:0x24])
+        payload = self._decode(response)
+
+        data = {}
+
+        if (payload):
+            data['payload'] = payload
+
+        return data
```
Why do we need this?

felipediel commented on broadlink/climate.py:
```diff
@@ -201,3 +201,252 @@ def set_schedule(self, weekday: List[dict], weekend: List[dict]) -> None:
+            health (bool):
+        """
+
+        args = locals()
+        args.pop('self')
```
Please do not use locals(), it makes the code difficult to maintain.

Enosh replied:
Is [this](https://github.com/Enosh/python-broadlink/blob/d885cc5d133c94f4201c6a2ce990021149a199a4/broadlink/climate.py#L393) a good way to do this?

felipediel commented on broadlink/climate.py:
```diff
@@ -201,3 +201,252 @@ def set_schedule(self, weekday: List[dict], weekend: List[dict]) -> None:
+        If not all parameters are specificed, will try to derive from the unit's current state. Doesn't work when powered down.
+        Args:
+            state (bool): power
+            set_temp (float): temperature set point 16<n<32
```
Rename set_temp to target_temp.

felipediel commented on broadlink/climate.py:
```diff
@@ -201,3 +201,252 @@ def set_schedule(self, weekday: List[dict], weekend: List[dict]) -> None:
+        """
+        
+        if (payload == 0):
+            packet = bytearray([0x0c, 0x00, 0xbb, 0x00, 0x06, 0x80, 0x00, 0x00, 0x02, 0x00, 0x21, 0x01, 0x1b, 0x7e])
+        elif (payload == 1):
+            packet = bytearray([0x0c, 0x00, 0xbb, 0x00, 0x06, 0x80, 0x00, 0x00, 0x02, 0x00, 0x11, 0x01, 0x2b, 0x7e])
+        elif (payload == 2):
+            packet = bytearray([0x0c, 0x00, 0xbb, 0x00, 0x06, 0x80, 0x00, 0x00, 0x02, 0x00, 0x41, 0x01, 0xfb, 0x7d])
```
This logic can be cleaned up with something like this:
```python3
        header = [0x0c, 0x00, 0xbb, 0x00, 0x06, 0x80, 0x00, 0x00, 0x02, 0x00]
        if (payload == 0):
            packet = header + bytes(0x21, 0x01, 0x1b, 0x7e])
        elif (payload == 1):
            packet = header + bytes([0x11, 0x01, 0x2b, 0x7e])
        elif (payload == 2):
            packet = header + bytes([0x41, 0x01, 0xfb, 0x7d])
```

felipediel commented on broadlink/climate.py:
```diff
@@ -201,3 +201,252 @@ def set_schedule(self, weekday: List[dict], weekend: List[dict]) -> None:
+
+        return data
+
+    def set_advanced(self, state:bool=None, mode:str=None, set_temp:float=None, speed:str=None, swing_v:str=None, swing_h:str=None,
+                        sleep:bool=None, display:bool=None, health:bool=None) -> bytes:
```
Please break this line and rename set_temp to target_temp.
```python3
def set_advanced(
    self,
    state: bool = None,
    mode: str = None,
    target_temp: float = None,
    speed: str = None,
    swing_v: str = None,
    swing_h: str = None,
    sleep: bool = None,
    display: bool = None,
    health: bool = None
) -> bytes:
```

felipediel commented on broadlink/climate.py:
```diff
@@ -201,3 +201,252 @@ def set_schedule(self, weekday: List[dict], weekend: List[dict]) -> None:
+        Args:
+            packet (list/bytearray/bytes): 
+        """
+        result = target - (sum([v if i % 2 == 0 else v << 8 for i, v in enumerate(packet)]) & 0xffff)
+        return (result & 0xff, (result >> 8) & 0xff)
```
Are you sure you are not overcomplicating things here? I haven't stopped to look at this logic carefully, but something tells me it could be simpler.

felipediel commented on broadlink/climate.py:
```diff
@@ -201,3 +201,252 @@ def set_schedule(self, weekday: List[dict], weekend: List[dict]) -> None:
+        response = self.send_packet(0x6a, bytearray(payload))
+        check_error(response[0x22:0x24])
+        response_payload = self._decode(response)
+        return response_payload
```
Add a line to the end of the file.

felipediel commented on broadlink/climate.py:
```diff
@@ -201,3 +201,252 @@ def set_schedule(self, weekday: List[dict], weekend: List[dict]) -> None:
+        assert(len(payload) == 32)
+        data = {}
+        data['state'] = True if (payload[0x14] & 0x20) == 0x20 else False
+        data['set_temp'] = (payload[0x0c] >> 3) + 8 + (0.0 if ((payload[0xe] & 0b10000000) == 0) else 0.5)
```
Rename set_temp to target_temp.

felipediel commented on broadlink/climate.py:
```diff
@@ -201,3 +201,252 @@ def set_schedule(self, weekday: List[dict], weekend: List[dict]) -> None:
+
+        return data
+
+    def get_ac_info(self, payload:bool=None) -> dict:
```
```python3
    def get_ac_info(self, payload: bool = None) -> dict:
```

felipediel commented:
Sorry for asking all this. You had already done a great job of cracking the codes. But if I change the code after we merge this I won't be able to test the changes because I don't have the device.

felipediel commented on broadlink/climate.py:
```diff
@@ -201,3 +201,252 @@ def set_schedule(self, weekday: List[dict], weekend: List[dict]) -> None:
+        data['display'] = (payload[0x16] & 0x10 == 0x10)
+        data['health'] = (payload[0x14] & 0b11 == 0b11)
+
+        checksum = self._calculate_checksum(payload[:0x19]) # checksum=(payload[0x1a] << 8) + payload[0x19]
```
How about something like this?
```python3
        checksum = resp[0x19] | (resp[0x1a] << 8)
        if sum(resp[:0x19], 0x20017) - sum(resp[0x19:0x1b]) & 0xffff != checksum:
            raise exception(-4008)  # Checksum error.
```
Are you sure 0x20017 is not an int (20017)? Please double check, this number does not feel right.

If so, you can use this:
```python3
        checksum = resp[0x19] | (resp[0x1a] << 8)
        if sum(resp[:0x19], 0x4e31) - sum(resp[0x19:0x1b]) & 0xffff != checksum:
            raise exception(-4008)  # Checksum error.
```

Enosh replied:
I don't fully understand how the checksum works, but I've done a lot of testing with this implementation without a problem showing up, will gladly send packet captures if you want to try. I'm pretty certain it's not a straight sum like your example.

> Sorry for asking all this. You had already done a great job of cracking the codes. But if I change the code after we merge this I won't be able to test the changes because I don't have the device.

It's actually nice to have someone looking over my code as I don't have a lot of experience.

felipediel commented on broadlink/climate.py:
```diff
@@ -288,38 +291,41 @@ def get_state(self, payload:bool=None) -> dict:
             data['speed'] = 'unrecognized value'
 
-        data['sleep'] = True if (payload[0x1a] & 0b100 == 0b000) else False
+        data['sleep'] = False if (payload[0x11] & 0b100 == 0b000) else True
```
```python3
        data['sleep'] = bool(payload[0x11] & 0b100)
```

felipediel commented on broadlink/climate.py:
```diff
@@ -201,3 +201,294 @@ def set_schedule(self, weekday: List[dict], weekend: List[dict]) -> None:
+                sleep (bool): 
+                display (bool):
+                health (bool):
+                cmnd_0d_rmask (int): unknown
+                cmnd_0e_rmask (int): unknown
+                cmnd_18 (int): unknown
```
What is this?

felipediel commented on broadlink/climate.py:
```diff
@@ -201,3 +201,294 @@ def set_schedule(self, weekday: List[dict], weekend: List[dict]) -> None:
+            sleep (bool) 
+            display (bool)
+            health (bool)
+            cmnd_0d_rmask (int): unknown
+            cmnd_0e_rmask (int): unknown
+            cmnd_18 (int): unknown
+        """
+        
+        received_state = self.get_state()
```
Something tells me that if we understand how this mask works we won't need to call self.get_state() before sending the command. We must do our best to keep communication atomic (we don't want to send more than one request per method).

thewh1teagle commented:
Hi guys
I have Tornado 12X SQ air conditioner
This should work on it?
How can i test it? 
Thank you so much!

KTibow commented:
@thewh1teagle download the repo, checkout the branch, build the wheel, and install with pip.

thewh1teagle commented:
> @thewh1teagle download the repo, checkout the branch, build the wheel, and install with pip.

Thanks. I used @Enosh [ Fork](https://github.com/Enosh/python-broadlink), I installed it with 
pip3 install . 
Then i used his [Tornado class.](https://github.com/Enosh/python-broadlink/blob/tornado_support/broadlink/climate.py#L205) and it works!!. 
I used his ```get_ac_info()``` method and i received ```{'state': True} ```
then i tried to use ```set_advanced()``` method like this
```dev.set_advanced(state=False,mode="auto", speed="low", target_temp=16)```
and It worked like a charm!

KTibow commented:
Mind fixing the conflicts @Enosh?

thewh1teagle commented:
Hi
Sometimes it says checksum failed. 
It works always after i use something from the app to control the ac, then i works once, and in the second try it says always checksum failed.
on Tornado 12X SQ air conditioner. 
Do you have any idea where can i start to check why i get that?

Enosh commented:
@felipediel `cmnd_0d_rmask`, `cmnd_0e_rmask` and `cmnd_18` are the parameters I haven't figured out (perhaps related to the unit's state and so ignored as input).
I removed from `set_advanced` the code retrieving the current state to fit the atomic communication model. And added `set_partial` to do this for user-friendliness.

I still don't know enough about checksums to do the checksum here right, so there are off by one errors from time to time.

@thewh1teagle `checksum fail ['xx', 'xx']` occurs only when getting the state from the unit and so shouldn't make any difference. It just means that the checksum calculation didn't work. If the unit has an issue with the packets being sent an exception is raised. 

I think I did a bit of a mess rebasing, hopefully it's fine.

thewh1teagle commented:
@Enosh Does heating control working on your ac? 
on mine completly not.

Enosh commented:
@thewh1teagle It does. I've just retested switching back and forth, changing speeds, swing modes, target temp and sleep.
Did you use the new `set_partial`? If so can you can add `print(' '.join('{:02x}'.format(x) for x in payload))` (prints the constructed payload in hex) at [line 507](https://github.com/Enosh/python-broadlink/blob/e16fe18dea170b8c0efc8211859e115ca49b9347/broadlink/climate.py#L507) and the results of `get_state(True)`.

thewh1teagle commented:
@Enosh 
```
python3 -i broadlink_cli --type 0x4e2a --host 192.168.1.100 --mac 23cfd6cd2ca5
```

### when i using mode=heating it does not controlling the ac
```
>>> dev.set_partial(state=False, mode="heating")
19 00 bb 00 06 80 00 00 0f 00 01 01 c0 e4 0d 20 40 80 00 00 00 00 00 00 05 1b f9 00 00 00 00 00
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/qlorg/.local/lib/python3.8/site-packages/broadlink/climate.py", line 562, in set_partial
    return self.set_advanced(**args)
  File "/home/qlorg/.local/lib/python3.8/site-packages/broadlink/climate.py", line 509, in set_advanced
    check_error(response[0x22:0x24])
  File "/home/qlorg/.local/lib/python3.8/site-packages/broadlink/exceptions.py", line 137, in check_error
    raise exception(error_code)
broadlink.exceptions.StorageError: [Errno -5] The device storage is full
```
### when i using mode=cooling it's working and i can hear 'beep' from the ac.
```
>>> dev.set_partial(state=False, mode="cooling")
19 00 bb 00 06 80 00 00 0f 00 01 01 c0 e4 0d 20 40 20 00 00 00 00 00 00 05 1b 59 00 00 00 00 00
b'\x0e\x00\xbb\x00\x07\x00\x00\x00\x04\x00\x01\x01\x1bY\x1d\xa5'
>>>
```

https://github.com/thewh1teagle/python-broadlink
This is the code i use. i merged your latest commits of course.

Enosh commented:
@thewh1teagle `state` has to be True i.e. powered on. I should probably add an error for that. The app doesn't allow you to change anything except display and settings when powered off.

thewh1teagle commented:
@Enosh 
No , it's not working when i set it to true in heating mode. again in cooling mode it working
```
>>> dev.set_partial(state=True, mode="heating")
19 00 bb 00 06 80 00 00 0f 00 01 01 c0 e4 0d 20 40 80 00 00 20 00 00 00 05 fb f8 00 00 00 00 00
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/qlorg/.local/lib/python3.8/site-packages/broadlink/climate.py", line 562, in set_partial
    return self.set_advanced(**args)
  File "/home/qlorg/.local/lib/python3.8/site-packages/broadlink/climate.py", line 509, in set_advanced
    check_error(response[0x22:0x24])
  File "/home/qlorg/.local/lib/python3.8/site-packages/broadlink/exceptions.py", line 137, in check_error
    raise exception(error_code)
broadlink.exceptions.StorageError: [Errno -5] The device storage is full
```
```
>>> dev.set_partial(state=True, mode="cooling")
19 00 bb 00 06 80 00 00 0f 00 01 01 c0 e4 0d 20 40 20 00 00 20 00 00 00 05 fb 58 00 00 00 00 00
b'\x0e\x00\xbb\x00\x07\x00\x00\x00\x04\x00\x01\x01\xfbX=\xa5'
>>>
```

Enosh commented:
@thewh1teagle I was able to reproduce, and _I think_ fix with this commit. Tell me if it works or you find anything else.

thewh1teagle commented:
@Enosh 
Thank you so much! now it's working!!
as i understand, it was only a logic issue, right?
another thing, i would like to know how did you manage to make all this, with a traffic sniffing? or maybe re the app? it's very interest me.

felipediel commented:
Hello @Enosh. Thanks for your dedication to this. And thanks for testing, @thewh1teagle. Tests are very welcome.

I will try to help with the masks and checksum. Could you provide me with some raw packets so I can start the investigation?

Enosh commented:
@felipediel Attached! [tornado-app-pcap.zip](https://github.com/mjg59/python-broadlink/files/5611255/tornado-app-pcap.zip)

@thewh1teagle Yeah an issue with the checksum logic.
After discovering that it was a Broadlink device, I found felipediel/broadlink-hacktools did a few packet captures going through the options in the app on my phone. Replayed the commands with this library, put them with the observed state in a spreadsheet and worked it out more or less.

felipediel commented on broadlink/climate.py:
```diff
@@ -238,3 +238,324 @@ def set_schedule(self, weekday: List[dict], weekend: List[dict]) -> None:
+
+        payload[0x0c] = ((int(target_temp) - 8 << 3) | swing_L)
+        payload[0x0d] = (int(swing_R) << 5) | cmnd_0d_rmask
+        payload[0x0e] = (0b10000000 if (target_temp % 1 == 0.5) else 0b0) | cmnd_0e_rmask
```
Let me see if I understand. Is this bit activated when the target temperature is a multiple of 0.5?

KTibow replied:
Nope, that's ending in .5

felipediel replied:
Thank you, I got it now. But still look like a magic behavior. Why this? I wish I could understand what is going on behind these bits.

felipediel replied:
Ah, now I see the full picture. You can only set target temperature with a 0.5 precision, correct? This bit indicates the presence of this decimal value.

Enosh replied:
> This bit indicates the presence of this decimal value.

Exactly.

felipediel replied:
We need to round the target temp before doing this to make this method more robust:
```python3
target_temp = round(target_temp * 2) / 2
```
If we don't round, 40.8 will set the temperature to 40, which is not the expected behavior.

Enosh replied:
Rounding is nice! haven't thought of that. Replaced the assert with a descriptive ValueError.

felipediel commented on broadlink/climate.py:
```diff
@@ -238,3 +238,324 @@ def set_schedule(self, weekday: List[dict], weekend: List[dict]) -> None:
+        elif (swing_h == 'ON'):
+            swing_R = 0b000
+        elif (swing_h >= 0 and swing_v <= 5):
+            swing_R = str(swing_v)
```
swing_h and swing_v are strings. You need to cast them for integer comparisons.

Just out of curiosity... Does swing_R = '0b110' do anything? And are you sure you are not mixing swing_h and swing_v in this part?

Enosh replied:
Added casting.
You're right I mixed which swing axis had the fixed positions.
`swing_R = '0b110'` gives a -5 error, since these positions are only available for the vertical.
This seems to have been the cause for the majority of the  issues I was having. Because the error is identical I thought it was an issue with the checksum, but it doesn't seem so.

felipediel commented on broadlink/climate.py:
```diff
@@ -238,3 +238,324 @@ def set_schedule(self, weekday: List[dict], weekend: List[dict]) -> None:
+        """Send a request for info from A/C unit and returns the response.
+        0 = GET_AC_INFO, 1 = GET_STATES, 2 = GET_SLEEP_INFO, 3 = unknown function
+        """
+        header = bytearray([0x0c, 0x00, 0xbb, 0x00, 0x06, 0x80, 0x00, 0x00, 0x02, 0x00])
```
I think [0x0c, 0x00] is the payload length.

felipediel replied:
The first two bytes carry the payload length. Three samples you gave me:

a) 0c 00 **bb 00 06 80 00 00 02 00 11 01 2b 7e** 00 00
b) 22 00 **bb 00 07 00 00 00 18 00 01 21 c0 24 00 00 00 3c 00 3e 00 00 00 00 00 00 00 00 00 00 00 00 00 07 64 38** 00 00 00 00 00 00 00 00 00 00 00 00
c) 19 00 **bb 00 07 00 00 00 0f 00 01 11 9f e4 2d 60 00 20 00 00 00 00 00 00 00 60 89** 00 00 00 00 00

We can use:
```
# For reading
payload_size = int.from_bytes(packet[:2], "little")

# For sending
payload_size = int.to_bytes(len(payload), 2, "little")
```
It reminds me that I have to rework the rm4 class because this is what happened there. In the new firmware the payload is prefixed with its length.

felipediel commented on broadlink/climate.py:
```diff
@@ -238,3 +238,324 @@ def set_schedule(self, weekday: List[dict], weekend: List[dict]) -> None:
+            speed_L, speed_R = 0x20, 0x00
+        elif (speed == 'mute'):
+            speed_L, speed_R = 0x40, 0x80
+            assert (mode == 'F')
```
Remove stale assert.

felipediel commented on broadlink/climate.py:
```diff
@@ -238,3 +238,324 @@ def set_schedule(self, weekday: List[dict], weekend: List[dict]) -> None:
+            mode_1 = 0x20
+        elif (mode == 'drying'):
+            mode_1 = 0x40
+            cmnd_0e_rmask = 0x16
```
This mask is probably eliminating the target temperature parameter here.
```python3
>>> bin(0x16)
'0b10110'
```

Enosh replied:
Works both with `cmnd_0e_rmask = 0x16` and `cmnd_0e_rmask = 0xd`, so I removed it. I think I added this because of errors that were actually related to swing_R.

Enosh commented:
After fixing the issues @felipediel noticed, everything I try now works, using the values `cmnd_0d_rmask = 0b100, cmnd_0e_rmask = 0b1101, cmnd_18 = 0b101`. The unit reports back different values in the response but no errors.
I've left overrides available, and made an optional parameter for get_state to report them.

I plan to move the parsing part out of get_state to reuse it for the responses in the other functions.

thewh1teagle commented on broadlink/climate.py:
```diff
@@ -546,16 +571,17 @@ def set_partial(self,
+            'health': health if health != None else received_state['health']
         }
 
-        if (args['mode'] == 'heating' or args['mode'] == 'fan' or args['mode'] == 'drying'):
```
Better to use
```python
if args['mode'] in ('heating', 'fan', 'drying'):
```

felipediel replied:
Looks better. A suggestion:
```python3
if args['mode'] in {'heating', 'fan', 'drying'}:
```

thewh1teagle replied:
I think that it's less pythonic and less effective comparing to tuple

felipediel replied:
We use sets when the order of the elements doesn't matter. They perform better in these situations and make intentions more explicit.

thewh1teagle commented on broadlink/climate.py:
```diff
@@ -546,16 +571,17 @@ def set_partial(self,
 
-        if (args['mode'] == 'heating' or args['mode'] == 'fan' or args['mode'] == 'drying'):
-            args['checksum_lbit'] = 1
-            if (args['swing_h'] == 'ON'):
```
No needs brackets in if statement in Python so you can remove them.

thewh1teagle commented on broadlink/climate.py:
```diff
@@ -460,34 +487,31 @@ def set_advanced(self,
-            speed_R = 0x40
             speed_L = 0x20 # doesn't matter
+            speed_R = 0x40
+            if not (mode == 'cooling' or mode == 'heating'):
```
A much more readable way to do it:
```python3
if mode not in ('cooling', 'heating'):
```

KTibow replied:
Any reason why tuples over lists?

thewh1teagle replied:
tuples are faster (I think)
https://stackoverflow.com/questions/1708510/list-vs-tuple-when-to-use-each

KTibow replied:
Thanks for teaching me that!

thewh1teagle commented on broadlink/climate.py:
```diff
@@ -362,23 +365,33 @@ def get_state(self, payload_debug: bool = None) -> dict:
+            checksum = list(self._calculate_checksum(payload[:0x19]))
+
+            # a kludge to make the checksum work
+            if (data['mode'] == 'heating' or data['mode'] == 'fan' or data['mode'] == 'drying'):
```
```python3
if data['mode'] in ('heating', 'fan', 'drying'):
```

felipediel commented on broadlink/__init__.py:
```diff
@@ -90,12 +90,12 @@
     0x60C8: (lb1, "LB1", "Broadlink"),
     0x6112: (lb1, "LB1", "Broadlink"),
     0x2722: (S1C, "S2KIT", "Broadlink"),
+    0X4E2A: (tornado, "TOP SQ X", "Tornado"),
```
We should consider naming the class with the model instead of the brand. Eventually Tornado will launch another device and we will have to create another class, and tornado_new is weird.

How about `top` or `sqx`? We need to make the type short so we can keep the code clean eg [here](https://github.com/home-assistant/core/blob/54425ae0f3c1e35c8040583edfe16ad3a1b7e243/homeassistant/components/broadlink/const.py#L8-L12).

Enosh replied:
I have a "SMART-16X SQ (WIFI)" (the numbers change based on size). Reading through the [catalogue](https://www.tornado-top.co.il/catalog/) there's also a TOP SQ WIFI __X series (which at some point I assumed were identical). Based on the app description both should work, but that is untested. So I renamed the class to `xsq`, does that work?

felipediel replied:
How does the model look like in the official app? If it looks specific (numbers included), the PIDs are different. If it looks generic, they are probably using the same PID for all these _X models.

Enosh replied:
It's utterly generic, the most specific it gets is "Air-conditioner".

felipediel replied:
@thewh1teagle What is your model?

felipediel replied:
I think we should use `sq` or `sq1`.

Enosh replied:
Sounds good, do you have a preference?

felipediel replied:
I think `sq1` will look better if we need to create a `sq2` in the future. We can use `sq1` for the class, `SQ1` for the type and `SQ` for the model name.

thewh1teagle replied:
@felipediel Sorry for late, my model is ```SMART-12X SQ (WIFI)```

felipediel replied:
Thank you!

felipediel commented:
I confirm that the checksum is matching the packets now 🎉

felipediel commented on broadlink/climate.py:
```diff
@@ -238,3 +238,393 @@ def set_schedule(self, weekday: List[dict], weekend: List[dict]) -> None:
+        else:
+            data['swing_v'] = 'unrecognized value'
+
+        mode = payload[0x11] >> 3 << 3
```
```python3
mode = payload[0x11] &~ 0b111
```

felipediel commented on broadlink/climate.py:
```diff
@@ -238,3 +238,393 @@ def set_schedule(self, weekday: List[dict], weekend: List[dict]) -> None:
+            health (bool)
+            clean (bool)
+            mildew (bool)
+            cmnd_0d_rmask (Optional[int]): override an unidentified option
+            cmnd_0e_rmask (Optional[int]): override an unidentified option
+            cmnd_18 (Optional[int]): override an unidentified option
+            payload_debug (Optional[bool]): print the constructed payload
```
These parameters can't go to production. We can use consts for now.

felipediel commented:
I think we should remove set_partial() and rename set_advanced() to set_state(), accepting a dictionary as input.

felipediel commented on broadlink/climate.py:
```diff
@@ -238,3 +238,393 @@ def set_schedule(self, weekday: List[dict], weekend: List[dict]) -> None:
+            True for verified success.
+        """
+
+        PREFIX = [0x19, 0x00, 0xbb, 0x00, 0x06, 0x80, 0x00, 0x00, 0x0f, 0x00,
+                  0x01, 0x01]  # 12B
+        MIDDLE = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+                  0, 0, 0, 0, 0]  # 13B + 2B checksum
+        SUFFIX = [0, 0, 0, 0, 0]  # 5B
+        payload = PREFIX + MIDDLE + SUFFIX
```
We could have a self._encode() for this, reference [here](https://github.com/mjg59/python-broadlink/blob/29345a129ff6c98962fccba8ada1ea049fd9ec16/broadlink/switch.py#L319-L330).

felipediel commented on broadlink/climate.py:
```diff
@@ -268,25 +268,40 @@ def _calculate_checksum(self, payload: bytes, byteorder: str) -> bytes:
+        packet.extend(self._calculate_checksum(packet[2:], 'little'))
+        packet.extend([0] * (packet_length - len(packet)))  # round size to 16
+
+        return packet
```
```python3
    def _encode(self, payload: bytes) -> bytes:
        """Encode the payload."""
        checksum = _calculate_checksum(payload, 'little')
        return struct.pack("H", len(payload) + 2) + payload + checksum
```

Enosh replied:
That's way more compact. Does anything else handle the padding?

felipediel commented on broadlink/climate.py:
```diff
@@ -238,3 +239,333 @@ def set_schedule(self, weekday: List[dict], weekend: List[dict]) -> None:
+        payload = self.decrypt(bytes(response[0x38:]))
+        return payload
+
+    def _calculate_checksum(self, payload: bytes, byteorder: str) -> bytes:
```
If we always return an integer, we can avoid the byteorder parameter.
```python3
    def _calculate_checksum(self, payload: bytes) -> int:
```

Enosh replied:
The get_ac_info request (and possibly others) return the checksum in the opposite byte order to other requests, I thought about having a default, but reckoned keeping it explicit would be more clear, no?

felipediel replied:
The idea is to have a [single responsibility](https://en.wikipedia.org/wiki/Single-responsibility_principle) for this method in order to make it more flexible/reusable. When we ask for a sum, we usually expect an integer. Packing the data for transportation is the responsibility of another layer.

felipediel commented on broadlink/climate.py:
```diff
@@ -238,3 +239,333 @@ def set_schedule(self, weekday: List[dict], weekend: List[dict]) -> None:
+
+        return data
+
+    def set_state(self, args: dict) -> bytes:
```
```python3
    def set_state(self, state: dict) -> bool:
```

felipediel commented on broadlink/climate.py:
```diff
@@ -238,3 +239,333 @@ def set_schedule(self, weekday: List[dict], weekend: List[dict]) -> None:
+            args = received_state
+            logging.debug("filled args %s", args)
+
+        PREFIX = [0xbb, 0x00, 0x06, 0x80, 0x00, 0x00, 0x0f, 0x00,
```
I think we can assume that all requests start with `bb 00 06 80` (and responses with `bb 00 07 00`). We can abstract this with self._encode() and self._decode(), respectively.

felipediel commented on broadlink/climate.py:
```diff
@@ -238,3 +239,333 @@ def set_schedule(self, weekday: List[dict], weekend: List[dict]) -> None:
+            0b111: 'OFF'
+        }.get(payload[0x0c] & 0b111, 'unrecognized value')
+
+        data['mode'] = {
+            0x00: 'auto',
+            0x20: 'cooling',
+            0x40: 'drying',
+            0x80: 'heating',
+            0xc0: 'fan'
+        }.get(payload[0x11] &~ 0b111, 'unrecognized value')  # noqa E225
```
My opinion on how to handle this like a boss. Import the IntEnum class and create child classes inside the controller class. So we would end up with something like this:
```python3
from enum import IntEnum, unique

class sq1:

    @unique
    class Mode(IntEnum):
        AUTO = 0
        COOLING = 0x20
        DRYING = 0x40
        HEATING = 0x80
        FAN = 0xc0
```

We can generate the mode in many different ways...
```python3
mode = self.Mode.FAN
mode = self.Mode["FAN"]
mode = self.Mode(0xc0)
```

We can use:
```python3
# For reading
mode = self.Mode(payload[0x11] &~ 0b111)

# For writing
packet[0xf] = mode << 3 | sleep << 2
```

This is like creating an interface for a normal integer.
```python3
# If we want to see the name
>>> mode.name
'COOLING'

# If we want to see the value
>>> mode.value
32
```
A ValueError is generated for invalid input:
```python3
>>> d = blk.hello("192.168.0.16")  # Example HVAC
>>> d.Mode(1000)
Traceback (most recent call last):
  File "<pyshell#59>", line 1, in <module>
    d.Mode(1000)
  File "/usr/lib/python3.6/enum.py", line 293, in __call__
    return cls.__new__(cls, value)
  File "/usr/lib/python3.6/enum.py", line 535, in __new__
    return cls._missing_(value)
  File "/usr/lib/python3.6/enum.py", line 548, in _missing_
    raise ValueError("%r is not a valid %s" % (value, cls.__name__))
ValueError: 1000 is not a valid Mode
```

External applications have access to an interface. They can see what options they have.
```python3
>>> d = blk.hello("192.168.0.16")  # Example HVAC
>>> [m for m in d.Mode.__members__]
['AUTO', 'COOLING', 'DRYING', 'HEATING', 'FAN']
```
Best of all: this is a normal integer that can be appended to the packet and serialized normally. What do you think?

Enosh replied:
I like it! Will implement.
Would you then pass a Mode object in the dictionary as input or the string representation?

felipediel commented on broadlink/climate.py:
```diff
@@ -238,3 +240,307 @@ def set_schedule(self, weekday: List[dict], weekend: List[dict]) -> None:
+        AUTO = 0xa0
+
+    @unique
+    class Swing_H(IntEnum):
```
Remove `_`

felipediel commented on broadlink/climate.py:
```diff
@@ -238,3 +240,307 @@ def set_schedule(self, weekday: List[dict], weekend: List[dict]) -> None:
+class sq1(device):
+    """Controls Tornado SMART X SQ series air conditioners."""
+
+    REQUEST_PREFIX = bytes([0xbb, 0x00, 0x06, 0x80])
+    RESPONSE_PREFIX = bytes([0xbb, 0x00, 0x07, 0x00])
```
Move REQUEST_PREFIX to self._encode and RESPONSE_PREFIX to self._decode. I think we can assume:
```python3
    REQUEST_PREFIX = bytes([0xbb, 0x00, 0x06, 0x80, 0x00, 0x00])
    RESPONSE_PREFIX = bytes([0xbb, 0x00, 0x07, 0x00, 0x00, 0x00])
```

felipediel commented on broadlink/climate.py:
```diff
@@ -238,3 +240,307 @@ def set_schedule(self, weekday: List[dict], weekend: List[dict]) -> None:
+        checksum = self._calculate_checksum(payload).to_bytes(2, 'little')
+        return (len(payload) + 2).to_bytes(2, 'little') + payload + checksum
+
+    def _send_short_payload(self, payload: int) -> bytes:
+        """Send a request for info from AC unit and returns the response.
+        0 = GET_AC_INFO, 1 = GET_STATES, 2 = GET_SLEEP_INFO
+        """
+        packet = bytes(
+            [0x00, 0x00, 0x02, 0x00]
+            + {
+                0: [0x21, 0x01],
+                1: [0x11, 0x01],
+                2: [0x41, 0x01]
+            }[payload]
+        )
+
+        response = self.send_packet(0x6a, self._encode(packet))
+        check_error(response[0x22:0x24])
+        return (self._decode(response))
```
```python3
    def _send_command(self, command: int, data: bytes = b'') -> bytes:
        """Send a command to the device."""
        packet = self._encode(command.to_bytes(2, "little") + data)
        response = self.send_packet(0x6a, packet)
        check_error(response[0x22:0x24])
        return self._decode(response)
```

felipediel commented on broadlink/climate.py:
```diff
@@ -238,3 +240,307 @@ def set_schedule(self, weekday: List[dict], weekend: List[dict]) -> None:
+                clean (bool):
+                mildew (bool):
+        """
+        payload = self._send_short_payload(1)
```
```python3
payload = self._send_command(0x141)
```

felipediel commented on broadlink/climate.py:
```diff
@@ -238,3 +240,307 @@ def set_schedule(self, weekday: List[dict], weekend: List[dict]) -> None:
+                state (bool): power
+                ambient_temp (float): ambient temperature
+        """
+        payload = self._send_short_payload(0)
```
```python3
payload = self._send_command(0x121)
```

felipediel commented on broadlink/climate.py:
```diff
@@ -238,3 +240,307 @@ def set_schedule(self, weekday: List[dict], weekend: List[dict]) -> None:
+        s = (s & 0xffff) + (s >> 16)
+        return (0xffff - s)
+
+    def _encode(self, payload: bytes) -> bytes:
+        """Encode payload (add length to beginning, checksum after payload)."""
+        payload = self.REQUEST_PREFIX + payload
+        checksum = self._calculate_checksum(payload).to_bytes(2, 'little')
+        return (len(payload) + 2).to_bytes(2, 'little') + payload + checksum
```
```python3
    def _encode(self, data: bytes) -> bytes:
        """Encode the data for transmission."""
        REQUEST_PREFIX = bytes([0xbb, 0x00, 0x06, 0x80, 0x00, 0x00])
        payload = REQUEST_PREFIX + len(data).to_bytes(2, "little") + data
        checksum = self._calculate_checksum(payload).to_bytes(2, 'little')
        return (len(payload) + 2).to_bytes(2, 'little') + payload + checksum
```

felipediel replied:
We can clean this up using struct. I'll see what I can do here.

felipediel replied:
```python3
    def _encode(self, data: bytes) -> bytes:
        """Encode data for transmission."""
        payload = struct.pack("HHHH", 0x00BB, 0x8060, 0x0000, len(data)) + data
        checksum = self._calculate_checksum(payload).to_bytes(2, 'little')
        return (len(payload) + 2).to_bytes(2, 'little') + payload + checksum
```

felipediel commented on broadlink/climate.py:
```diff
@@ -238,3 +240,307 @@ def set_schedule(self, weekday: List[dict], weekend: List[dict]) -> None:
+
+        speed_L = self.Speed(args['speed']).value
+
+        payload = self._encode(bytes(
+            [
+                0x00,
+                0x00,
+                0x0f,
+                0x00,
+                0x01,
+                0x01,
+                (int(args['target_temp']) - 8 << 3) | swing_L,
+                (swing_R << 5) | CMND_0B_RMASK,
+                (0b10000000 if (args['target_temp'] % 1 == 0.5) else 0
+                 | CMND_0C_RMASK),
+                speed_L,
+                speed_R,
+                mode | (0b100 if args['sleep'] else 0b000),
+                0x00,
+                0x00,
+                (0b100000 if args['state'] else 0b000000
+                 | 0b100 if args['clean'] else 0b000
+                 | 0b11 if args['health'] else 0b00),
+                0x00,
+                (0b10000 if args['display'] else 0b00000
+                 | 0b1000 if args['mildew'] else 0b0000),
+                0x00,
+                CMND_16
+            ]
+        ))
```
```python3
        data = bytes(
            [
                (int(args['target_temp']) - 8 << 3) | swing_L,
                (swing_R << 5) | CMND_0B_RMASK,
                (0b10000000 if (args['target_temp'] % 1 == 0.5) else 0
                 | CMND_0C_RMASK),
                speed_L,
                speed_R,
                mode | (0b100 if args['sleep'] else 0b000),
                0x00,
                0x00,
                (0b100000 if args['state'] else 0b000000
                 | 0b100 if args['clean'] else 0b000
                 | 0b11 if args['health'] else 0b00),
                0x00,
                (0b10000 if args['display'] else 0b00000
                 | 0b1000 if args['mildew'] else 0b0000),
                0x00,
                CMND_16
            ]
        ))
```

felipediel commented on broadlink/climate.py:
```diff
@@ -238,3 +240,307 @@ def set_schedule(self, weekday: List[dict], weekend: List[dict]) -> None:
+        ))
+        logging.debug("Constructed payload:\n%s", payload.hex(' '))
+
+        response = self.send_packet(0x6a, payload)
+        check_error(response[0x22:0x24])
+        response_payload = self._decode(response)
```
```python3
response_payload = self._send_command(0x101, data)
```

felipediel commented on broadlink/climate.py:
```diff
@@ -238,3 +240,307 @@ def set_schedule(self, weekday: List[dict], weekend: List[dict]) -> None:
+            raise ValueError(f"target_temp out of range, value: {args['target_temp']}")  # noqa E501
+
+        # Creating a new instance verifies the type
+        swing_R = self.Swing_H(args['swing_h']).value
+        swing_L = self.Swing_V(args['swing_v']).value
+
+        mode = self.Mode(args['mode']).value
```
We can remove `.value` here. The result will always be an integer (with a cool interface).

felipediel commented on broadlink/climate.py:
```diff
@@ -238,3 +240,307 @@ def set_schedule(self, weekday: List[dict], weekend: List[dict]) -> None:
+        """Set parameters of unit.
+
+        Args:
+            args (dict): if any are missing the current value will be retrived
+                state (bool): power
```
I think we should rename state (bool) to power and args (dict) to state.

felipediel commented:
Hi @Enosh. Thanks again for this. I will solve conflicts and create a PR to the dev branch, we will continue from there, ok?

felipediel commented:
@Enosh I couldn't leave all of our effort behind, so I rebased and created a PR on the dev branch: https://github.com/mjg59/python-broadlink/pull/520. After I make some improvements, could you please help me to test?

thewh1teagle commented:
Any update on this? does the library supports tornado ACs?

felipediel commented:
Closed with https://github.com/mjg59/python-broadlink/pull/520. Thanks a lot @enosh!
